google/cloud/storage/blob.py

```diff
@@ -2162,14 +2189,18 @@ def _do_upload(
                   **only** response in the multipart case and it will be the
                   **final** response in the resumable case.
         """
-        if if_metageneration_match is None and num_retries is None:
-            # Uploads are only idempotent (safe to retry) if
-            # if_metageneration_match is set. If it is not set, the default
-            # num_retries should be 0. Note: Because retry logic for uploads is
-            # provided by the google-resumable-media-python package, it doesn't
-            # use the ConditionalRetryStrategy class used in other API calls in
-            # this library to solve this problem.
-            num_retries = 0
+
+        # Handle ConditionalRetryPolicy.
+        if isinstance(retry, ConditionalRetryPolicy):
+            # Conditional retries are designed for non-media calls, which change
+            # arguments into query_params dictionaries. Media operations work
+            # differently, so here we make a "fake" query_params to feed to the
+            # ConditionalRetryPolicy.
+            query_params = {
```

Comment:
If `if_generation_match` and `if_metageneration_match` are not set, will the values in this dict be 0 or `None`? I'm a little confused because the docstring indicates that the param is optional but the type is  listed as`long`.

Reply:
They will be None. Optional in the docstring means it may be either the declared type or None.